File: app/src/main/java/com/arjanvlek/oxygenupdater/Model/UpdateData.java
package com.arjanvlek.oxygenupdater.Model;

import com.arjanvlek.oxygenupdater.Support.SettingsManager;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;

@JsonIgnoreProperties(ignoreUnknown = true)
public class UpdateData {

    private Long id;
    private String versionNumber;
    private String otaVersionNumber;
    private String description;
    private String downloadUrl;
    private int downloadSize;
    private String filename;
    private String MD5Sum;
    private String information;
    private boolean updateInformationAvailable;
    private boolean systemIsUpToDate;


    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getVersionNumber() {
        return versionNumber;
    }

    @JsonProperty("version_number")
    public void setVersionNumber(String versionNumber) {
        this.versionNumber = versionNumber;
    }


    public String getOtaVersionNumber() {
        return otaVersionNumber;
    }

    @JsonProperty("ota_version_number")
    public void setOtaVersionNumber(String otaVersionNumber) {
        this.otaVersionNumber = otaVersionNumber;
    }


    public String getDescription() {
        return description;
    }


    public void setDescription(String description) {
        this.description = description;
    }


    public String getDownloadUrl() {
        return downloadUrl;
    }

    @JsonProperty("download_url")
    public void setDownloadUrl(String downloadUrl) {
        this.downloadUrl = downloadUrl;
    }


    public int getDownloadSize() {
        return (downloadSize / 1048576 );
    }

    @JsonProperty("download_size")
    public void setDownloadSize(int downloadSize) {
        this.downloadSize = downloadSize;
    }


    public String getFilename() {
        return filename;
    }

    public void setFilename(String filename) {
        this.filename = filename;
    }


    public String getMD5Sum() {
        return MD5Sum;
    }

    @JsonProperty("md5sum")
    public void setMD5Sum(String MD5Sum) {
        this.MD5Sum = MD5Sum;
    }


    public String getInformation() {
        return information;
    }

    public void setInformation(String information) {
        this.information = information;
    }

    public boolean isUpdateInformationAvailable() {
        return updateInformationAvailable || (versionNumber != null);
    }

    @JsonProperty("update_information_available")
    public void setUpdateInformationAvailable(boolean updateInformationAvailable) {
        this.updateInformationAvailable = updateInformationAvailable;
    }

    public boolean isSystemIsUpToDateCheck(SettingsManager settingsManager) {
        //noinspection SimplifiableIfStatement
        if(settingsManager != null && settingsManager.getPreference(SettingsManager.PROPERTY_SHOW_IF_SYSTEM_IS_UP_TO_DATE, true)) {
            return systemIsUpToDate;
        } else {
            return false;
        }
    }

    public boolean isSystemIsUpToDate() {
        return systemIsUpToDate;
    }

    @JsonProperty("system_is_up_to_date")
    public void setSystemIsUpToDate(boolean systemIsUpToDate) {
        this.systemIsUpToDate = systemIsUpToDate;
    }

 }

File: app/src/main/java/com/arjanvlek/oxygenupdater/models/UpdateData.kt
package com.arjanvlek.oxygenupdater.models

import android.os.Parcel
import android.os.Parcelable
import android.util.SparseBooleanArray
import com.arjanvlek.oxygenupdater.internal.settings.SettingsManager
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonProperty

@JsonIgnoreProperties(ignoreUnknown = true)
data class UpdateData(
    var id: Long? = null,
    var versionNumber: String? = null,
    var otaVersionNumber: String? = null,
    var description: String? = null,
    var downloadUrl: String? = null,
    var downloadSize: Long = 0,
    var filename: String? = null,

    @JsonProperty("md5sum")
    var mD5Sum: String? = null,

    var information: String? = null,
    var updateInformationAvailable: Boolean = false,

    var systemIsUpToDate: Boolean = false
) : Parcelable, FormattableUpdateData {

    val downloadSizeInMegabytes = downloadSize / 1048576L

    val isUpdateInformationAvailable = updateInformationAvailable || versionNumber != null

    // Formatting library: interface FormattableUpdateData
    override val internalVersionNumber = versionNumber
    override val updateDescription = description

    @JsonProperty("update_information_available")
    fun setIsUpdateInformationAvailable(updateInformationAvailable: Boolean) {
        this.updateInformationAvailable = updateInformationAvailable
    }

    fun isSystemIsUpToDateCheck(settingsManager: SettingsManager?) = if (settingsManager != null
        && settingsManager.getPreference(SettingsManager.PROPERTY_ADVANCED_MODE, false)
    ) false
    else systemIsUpToDate

    override fun describeContents() = 0

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeLong(id ?: -1L)
        dest.writeString(versionNumber)
        dest.writeString(otaVersionNumber)
        dest.writeString(description)
        dest.writeString(downloadUrl)
        dest.writeLong(downloadSize)
        dest.writeString(filename)
        dest.writeString(mD5Sum)
        dest.writeString(information)

        SparseBooleanArray().apply {
            put(0, updateInformationAvailable)
            put(1, systemIsUpToDate)

            dest.writeSparseBooleanArray(this)
        }
    }

    companion object {
        val CREATOR = object : Parcelable.Creator<UpdateData> {
            override fun createFromParcel(parcel: Parcel): UpdateData? {
                val data = UpdateData(
                    id = parcel.readLong(),
                    versionNumber = parcel.readString(),
                    otaVersionNumber = parcel.readString(),
                    description = parcel.readString(),
                    downloadUrl = parcel.readString(),
                    downloadSize = parcel.readLong(),
                    filename = parcel.readString(),
                    mD5Sum = parcel.readString(),
                    information = parcel.readString()
                )

                parcel.readSparseBooleanArray()?.let {
                    data.setIsUpdateInformationAvailable(it[0])
                    data.systemIsUpToDate = it[1]
                }

                return data
            }

            override fun newArray(size: Int) = arrayOfNulls<UpdateData?>(size)
        }
    }
}
